src/Sylius/Bundle/AdminBundle/Action/ShipShipmentOnIndexPageAction.php

```diff
@@ -0,0 +1,102 @@
+use SM\Factory\FactoryInterface as StateMachineFactoryInterface;
+use Sylius\Component\Core\Model\ShipmentInterface;
+use Sylius\Component\Core\Repository\ShipmentRepositoryInterface;
+use Sylius\Component\Resource\Exception\UpdateHandlingException;
+use Sylius\Component\Shipping\ShipmentTransitions;
+use Symfony\Component\EventDispatcher\Debug\TraceableEventDispatcher;
+use Symfony\Component\EventDispatcher\GenericEvent;
+use Symfony\Component\HttpFoundation\RedirectResponse;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\HttpFoundation\Session\Session;
+use Symfony\Component\HttpKernel\Exception\HttpException;
+use Symfony\Component\Security\Csrf\CsrfToken;
+use Symfony\Component\Security\Csrf\CsrfTokenManagerInterface;
+use Webmozart\Assert\Assert;
+
+final class ShipShipmentOnIndexPageAction
+{
+    /** @var CsrfTokenManagerInterface */
+    private $csrfTokenManager;
+
+    /** @var ShipmentRepositoryInterface */
+    private $shipmentRepository;
+
+    /** @var StateMachineFactoryInterface */
+    private $stateMachineFactory;
+
+    /** @var ObjectManager */
+    private $shipmentManager;
+
+    /** @var TraceableEventDispatcher */
+    private $eventDispatcher;
+
+    public function __construct(
+        CsrfTokenManagerInterface $csrfTokenManager,
+        ShipmentRepositoryInterface $shipmentRepository,
+        StateMachineFactoryInterface $stateMachineFactory,
+        ObjectManager $shipmentManager,
+        TraceableEventDispatcher $eventDispatcher
+    ) {
+        $this->csrfTokenManager = $csrfTokenManager;
+        $this->shipmentRepository = $shipmentRepository;
+        $this->stateMachineFactory = $stateMachineFactory;
+        $this->shipmentManager = $shipmentManager;
+        $this->eventDispatcher = $eventDispatcher;
+    }
+
+    public function __invoke(Request $request): Response
+    {
+        $shipmentId = $request->attributes->get('id');
+        Assert::notNull($shipmentId);
+
+        /** @var ShipmentInterface $shipment */
+        $shipment = $this->shipmentRepository->find($shipmentId);
+        Assert::notNull($shipment);
+
+        if (!$this->csrfTokenManager->isTokenValid(new CsrfToken($shipmentId, $request->request->get('_csrf_token')))) {
+            throw new HttpException(Response::HTTP_FORBIDDEN, 'Invalid csrf token.');
+        }
+
+        $stateMachine = $this->stateMachineFactory->get($shipment, ShipmentTransitions::GRAPH);
+
+        if (!$stateMachine->can(ShipmentTransitions::TRANSITION_SHIP)) {
+            throw new UpdateHandlingException();
+        }
+
+        $trackingCode = $request->request->get('sylius_shipment_ship_tracking');
+        if ($trackingCode !== null) {
+            $shipment->setTracking($trackingCode);
+        }
+        $this->stateMachineFactory
+            ->get($shipment, ShipmentTransitions::GRAPH)
+            ->apply(ShipmentTransitions::TRANSITION_SHIP)
+        ;
```

Comment:
I feel a little bit uncomfortable that there is so many logic in the controller action - I know it's used only there, but still, it's not quite SOLID 😄